README.md
# YandexCloud

A simple Elixir wrapper to Yandex Cloud API

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `yandex_cloud` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:yandex_cloud, "~> 0.3.2"}
  ]
end
```

## Get access to API

### Getting a Cloud API KEY

To make api requests in Yandex.Cloud you must include IAM-token in request headers and for getting IAM-token you must have Oauth token.

Instruction for getting Oauth token is [here](https://cloud.yandex.com/docs/iam/operations/iam-token/create)

### Configuration

The default behaviour is to configure using the application environment:

In config/config.exs, add:

```elixir
config :yandex_cloud, cloud_api_key: "API_KEY"
```

## Usage

## Get IAM-token for cloud API access

Request for getting IAM-token for access to cloud API. Valid 12 hours.

```elixir
# without params
YandexCloud.get_iam_token

# or with key
YandexCloud.get_iam_token(%{oauth_token: "API_KEY"})
```

#### Options

    oauth_token - API KEY, optional

#### Responses

```elixir
# successful response
{:ok, %{"iamToken" => iam_token}}

# response with errors
{:error, %{"code" => code, "message" => message}}
```

## Translate service

### Configuration

For using Translate service you need additionally add FOLDER_ID.

Instruction for getting FOLDER_ID is [here](https://cloud.yandex.com/docs/translate/concepts/auth)

In config/config.exs, add:

```elixir
config :yandex_cloud, translate_folder_id: "FOLDER_ID"
```

Or you can add folder_id param to each request.

### Supported languages

Request for getting list of supported languages is #languages.

```elixir
YandexCloud.Translate.languages(%{iam_token: ""})
```

#### Options

    iam_token - IAM-token, required
    folder_id - folder ID of your account at Yandex.Cloud, optional

### Detection

Request for detecting language of text is #detect.

```elixir
YandexCloud.Translate.detect(%{iam_token: "", text: "Hello"})
```

#### Options

    iam_token - IAM-token, required
    folder_id - folder ID of your account at Yandex.Cloud, optional
    text - text for detection, required
    hint - list of possible languages, optional, example - "en,ru"

### Translation

Request for translating text is #translate.

```elixir
YandexCloud.Translate.translate(%{iam_token: "", text: "Hello", target: "ru"})
```

#### Options

    iam_token - IAM-token, required
    folder_id - folder ID of your account at Yandex.Cloud, optional
    text - text for detection, required
    source - source language, ISO 639-1 format (like "en"), optional
    target - target language, ISO 639-1 format (like "ru"), required
    format - text format, one of the [plain|html], default - plain, optional

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/kortirso/ex_yandex_cloud.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).

## Disclaimer

Use this package at your own peril and risk.

## Documentation

Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
be found at [https://hexdocs.pm/yandex_translator](https://hexdocs.pm/yandex_cloud).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.3.2] - 2019-08-14
### Added
- add mocking of requests in tests

### Modified
- change attrs from list to map

## [0.3.1] - 2019-02-01
### Modified
- rename langs -> languages

## [0.3.0] - 2019-02-01
### Added
- requests to Translate service
- tests

### Modified
- README

## [0.2.0] - 2019-02-01
### Added
- YandexCloud.Auth module with token() function for getting IAM-token
- tests

### Modified
- README


test/yandex_cloud_test.exs
defmodule YandexCloudTest do
  use ExUnit.Case
end


test/yandex_cloud/auth_test.exs
defmodule YandexCloud.AuthTest do
  use ExUnit.Case
  alias YandexCloud.Auth

  test "request for getting iam_token with invalid key" do
    {:error, %{"code" => code, "message" => message}} = Auth.token(%{oauth_token: "12345"})

    assert code == 16
    assert message == "Token is invalid or has expired."
  end

  test "request for getting iam_token with valid key" do
    {:ok, %{"iamToken" => iam_token}} = Auth.token()

    assert is_binary(iam_token) == true
  end
end


test/yandex_cloud/translate_test.exs
defmodule YandexCloud.TranslateTest do
  use ExUnit.Case
  import Mock
  alias YandexCloud.Translate

  setup_all do
    {:ok, %{"iamToken" => iam_token}} = YandexCloud.get_iam_token
    {:ok, iam_token: iam_token}
  end

  test "request for getting languages with invalid key" do
    {:error, %{"error_message" => error_message}} = Translate.languages(%{iam_token: "1"})

    assert error_message == "rpc error: code = Unauthenticated desc = The token is invalid"
  end

  test "request for detecting language with invalid key" do
    {:error, %{"error_message" => error_message}} = Translate.detect(%{iam_token: "1", text: "Hello"})

    assert error_message == "rpc error: code = Unauthenticated desc = The token is invalid"
  end

  test "request for translating text with invalid key" do
    {:error, %{"error_message" => error_message}} = Translate.translate(%{iam_token: "1", text: "Hello", lang: "ru"})

    assert error_message == "rpc error: code = Unauthenticated desc = The token is invalid"
  end

  test "request for getting languages", state do
    with_mock Translate, [languages: fn(_url) -> {:ok, %{"languages" => []}} end] do
      {:ok, %{"languages" => languages}} = Translate.languages(%{iam_token: state[:iam_token]})

      assert is_list(languages) == true
    end
  end

  test "request for detecting language", state do
    with_mock Translate, [detect: fn(_url) -> {:ok, %{"language" => "en"}} end] do
      {:ok, %{"language" => language}} = Translate.detect(%{iam_token: state[:iam_token], text: "Hello"})

      assert language == "en"
    end
  end

  test "request for translating text", state do
    with_mock Translate, [translate: fn(_url) -> {:ok, %{"translations" => [%{"text" => "Saludar"}]}} end] do
      {:ok, %{"translations" => [%{"text" => text}]}} = Translate.translate(%{iam_token: state[:iam_token], text: "Hello", target: "es"})

      assert text == "Saludar"
    end
  end
end


mix.exs
defmodule YandexCloud.MixProject do
  use Mix.Project

  @description """
  API Wrapper for Yandex Cloud services with Elixir
  """

  def project do
    [
      app: :yandex_cloud,
      version: "0.3.2",
      elixir: "~> 1.7",
      description: @description,
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      package: package(),
      deps: deps(),
      source_url: "https://github.com/kortirso/ex_yandex_cloud"
    ]
  end

  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp deps do
    [
      {:httpoison, "~> 1.4"},
      {:poison, "~> 3.1.0"},
      {:mock, "~> 0.3.0", only: :test},
      {:ex_doc, "~> 0.19", only: :dev}
    ]
  end

  defp package do
    [
      maintainers: ["Bogdanov Anton"],
      licenses: ["MIT"],
      links: %{"github" => "https://github.com/kortirso/ex_yandex_cloud"}
    ]
  end
end


lib/yandex_cloud/auth.ex
defmodule YandexCloud.Auth do
  @moduledoc """
  Auth requests
  """

  @iam_token_url "https://iam.api.cloud.yandex.net/iam/v1/tokens"
  @iam_token_headers [{"Content-Type", "application/json"}]

  @doc """
  Get IAM-token for using it in requests to YandexCloud. Valid 12 hours.

  ## Example

      iex> YandexCloud.Auth.token
      {:ok, %{"iamToken" => ""}}

  """
  @spec token(none()) :: {:ok, %{iamToken: String.t()}}

  def token, do: token(%{oauth_token: nil})

  @doc """
  Get IAM-token for using it in requests to YandexCloud. Valid 12 hours.

  ## Examples

      iex> YandexCloud.Auth.token(%{oauth_token: oauth_token})
      {:ok, %{"iamToken" => ""}}

  """
  @spec token(map()) :: {:ok, %{iamToken: String.t()}}

  def token(%{oauth_token: oauth_token}) do
    oauth_token
    |> fetch_iam_token()
    |> parse()
  end

  defp fetch_iam_token(api_key) do
    body = Poison.encode!(%{"yandexPassportOauthToken" => cloud_api_key(api_key)})

    case HTTPoison.post(@iam_token_url, body, @iam_token_headers) do
      {:ok, %HTTPoison.Response{status_code: 200, body: body}} -> {:ok, body}
      {:ok, %HTTPoison.Response{body: body}} -> {:error, body}
      {:error, %HTTPoison.Error{reason: reason}} -> {:error, reason}
    end
  end

  defp cloud_api_key(nil), do: Application.get_env(:yandex_cloud, :cloud_api_key) || ""
  defp cloud_api_key(api_key), do: api_key

  # parse results
  defp parse({result, response}), do: {result, Poison.Parser.parse!(response)}
end


lib/yandex_cloud/translate.ex
defmodule YandexCloud.Translate do
  @moduledoc """
  Client requests to Translate service
  """

  @cloud_base_url "https://translate.api.cloud.yandex.net/translate/v1/"

  @doc """
  Get available languages for translation.

  ## Example

      iex> YandexCloud.Translate.languages(%{iam_token: ""})
      {:ok, %{"languages" => [%{"language" => "az"}, %{...}, ...]}}

  ### Options

      iam_token - IAM-token, required
      folder_id - folder ID of your account at YandexCloud, optional

  """
  @spec languages(map()) :: {:ok, %{}}

  def languages(%{} = options), do: call("languages", options)

  @doc """
  Detect language for text.

  ## Example

      iex> YandexCloud.Translate.detect(%{iam_token: "", text: "Hello"})
      {:ok, %{"language" => "en"}}

  ### Options

      iam_token - IAM-token, required
      folder_id - folder ID of your account at YandexCloud, otional]
      text - text for detection, required
      hint - list of possible languages, optional, example - "en,ru"

  """
  @spec detect(map()) :: {:ok, %{}}

  def detect(%{} = options), do: call("detect", options)

  @doc """
  Translate word or phrase.

  ## Example

      iex> YandexCloud.Translate.translate(%{iam_token: iam_token, text: "hello world", source: "en", target: "es"})
      {:ok, %{"translations" => [%{"text" => "hola mundo"}]}}

  ### Options

      iam_token - IAM-token, required
      folder_id - folder ID of your account at YandexCloud, optional
      text - text for detection, required
      source - source language, ISO 639-1 format (like "en"), optional
      target - target language, ISO 639-1 format (like "ru"), required
      format - text format, one of the [plain|html], default - plain, optional

  """
  @spec translate(map()) :: {:ok, %{}}

  def translate(%{} = options), do: call("translate", options)

  # perform request
  defp call(type, args) do
    type
    |> fetch(args)
    |> parse()
  end

  # MAIN FUNCTIONS

  # make request
  defp fetch(type, args) do
    case HTTPoison.post(@cloud_base_url <> type, body(args, type), headers(args[:iam_token])) do
      {:ok, %HTTPoison.Response{status_code: 200, body: body}} -> {:ok, body}
      {:ok, %HTTPoison.Response{body: body}} -> {:error, body}
      {:error, %HTTPoison.Error{reason: reason}} -> {:error, reason}
    end
  end

  # parse results
  defp parse({result, response}), do: {result, Poison.Parser.parse!(response)}

  # ADDITIONAL FUNCTIONS

  # generate body for request
  defp body(args, type) do
    args
    |> Map.merge(%{folderId: cloud_folder_id(args[:folder_id])})
    |> Stream.filter(fn {key, _} -> Enum.member?(valid_args(type), key) end)
    |> Stream.map(fn {key, value} -> "#{key}=#{URI.encode_www_form(value)}" end)
    |> Enum.join("&")
    #args
    #|> Keyword.put(:folderId, cloud_folder_id(args[:folder_id]))
    #|> Stream.filter(fn {key, _} -> Enum.member?(valid_args(type), key) end)
    #|> Stream.map(fn {key, value} -> "#{key}=#{URI.encode_www_form(value)}" end)
    #|> Enum.join("&")
  end

  # list with available options based on request type
  defp valid_args(type) do
    case type do
      "languages" -> [:folderId]
      "detect" -> [:folderId, :text, :hint]
      "translate" -> [:folderId, :text, :source, :target, :format]
      _ -> []
    end
  end

  defp cloud_folder_id(nil), do: Application.get_env(:yandex_cloud, :translate_folder_id) || ""
  defp cloud_folder_id(folder_id), do: folder_id

  # define headers for request
  defp headers(iam_token), do: [{"Content-Type", "application/x-www-form-urlencoded"}, {"Authorization", "Bearer #{iam_token}"}]
end
